### PR TITLE
OSASINFRA-3583: openstack: singlestackv6 job

### PR DESCRIPTION
**Pull request by EmilienM**
A new CI job that will test the deployment of OpenShift on top of
OpenStack in a IPv6 disconnected environment.

* Create the optional pre-submit jobs in the installer and MCO (starting
  4.18) and a periodic job for shiftstack.
* Create a new provision step called `openstack-provision-mirror` which
  is in charge of creating a VM in a disconnected network and configure
  docker registry v2 on it. You'll notice that because the network is
  disconnected, we SSH the VM through the proxy. Also, there is a step
  to deprovision the mirror so we can collect the registry logs when
  needed.
* Create a new chain: `ipi-openstack-pre-disconnected`.
  Because we need to configure RBC before populating the mirror,
  we now have a custom chain for disconnected environments which also
  calls existing refs for mirroring a release payload into a
  disconnected registry and then populate the install-config.yaml.
  We didn't have to create steps for that, we re-use the ones from
  OpenShift QE.
* Create a new workflow: `openshift-e2e-openstack-singlestackv6` which
  contains the right steps and environment variables for that kind
  of deployment.
* In `openstack-conf-clouds`, we now duplicate the cloud entry in
  clouds.yaml to create another entry for ipv6.
  This entry will be used in install-config so when the cluster is being
  deployed, OpenShift will talk to the OpenStack IPv6 endpoint.
* In `openstack-conf-installconfig`, we now support a new CONFIG_TYPE
  for singlestackv6, where we configure the IPv6 network and IPs for API
  & Ingress. Note that we use the ipv6 entry from clouds.yaml.
* In `openstack-provision-dnsrecords`, we now configure the DNS record
  for the mirror-registry when needed.

### CODE REVIEW

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6
let's revert the last commit if it didn't mirror the registry

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
I move the mirror code into a separate PR: https://github.com/openshift/release/pull/56465

**Comment by EmilienM** [comment]
this is ready for review otherwise

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

File: ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+if ! retry 60 5 ssh_via_proxy "uname -a"; then
+		echo "ERROR: Bastion proxy is not reachable via $mirror_ipv4 - check logs:"
+    openstack console log show ${server_id}
+		exit 1
+fi
+
+MIRROR_REGISTRY_DNS_NAME="mirror-registry.${CLUSTER_NAME}.${BASE_DOMAIN}"
+MIRROR_REGISTRY_CREDENTIALS=$(<"/var/run/vault/mirror-registry/registry_creds")
```
Comment (MaysaMacedo) [inline ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh:105]: where this credential comes from?

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Review by MaysaMacedo — COMMENTED** [review]
will the OCP cluster be able to resolve the DNS over the IPv6 network?
I thought the bastion would need to have a DNS server

File: ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh (hunk 1 above)
Comment (EmilienM) [reply ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh:105]: from OCP CI vault, I'm re-using the same credentials as OpenShift QE for simplicity.

**Comment by EmilienM** [comment]
> will the OCP cluster be able to resolve the DNS over the IPv6 network? I thought the bastion would need to have a DNS server

no, I've changed `openstack-provision-dnsrecords` to support the mirror DNS record :)

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

File: ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+>&2 echo "Created necessary security group rules in ${sg_id}"
+
+server_params="--network $CONTROL_PLANE_NETWORK --image $BASTION_IMAGE --flavor $BASTION_FLAVOR \
+  --security-group $sg_id --key-name mirror-${CLUSTER_NAME}-${CONFIG_TYPE}"
+
+server_id="$(openstack server create -f value -c id $server_params \
+		"mirror-$CLUSTER_NAME-${CONFIG_TYPE}")"
+>&2 echo "Created nova server mirror-${CLUSTER_NAME}-${CONFIG_TYPE}: ${server_id}"
+
+sleep 10
```
Comment (MaysaMacedo) [inline ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh:64]: do we need this sleep here? is it better to add a --wait?

File: ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+>&2 cat << EOF > $WORK_DIR/deploy_mirror.sh
+#!/usr/bin/env bash
+set -e
+sudo mkdir -p /opt/registry/{auth,certs,data}
+sudo openssl req -newkey rsa:4096 -nodes -sha256 -keyout /opt/registry/certs/domain.key -x509 -days 1 -subj "/CN=mirror-$CLUSTER_NAME-${CONFIG_TYPE}" -addext "subjectAltName=DNS:$MIRROR_REGISTRY_DNS_NAME,DNS:mirror-$CLUSTER_NAME-${CONFIG_TYPE}" -out /opt/registry/certs/domain.crt
+sudo cp /opt/registry/certs/domain.crt /etc/pki/ca-trust/source/anchors/domain.crt   
+sudo mv /tmp/htpasswd /opt/registry/auth/htpasswd
+sudo update-ca-trust
+sudo dnf install -y podman
+sudo podman create --name registry -p 5000:5000 --net host \
```
Comment (MaysaMacedo) [inline ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh:118]: There is a tool available to create a registry https://docs.openshift.com/container-platform/4.16/installing/disconnected_install/installing-mirroring-creating-registry.html#mirror-registry-localhost_installing-mirroring-creating-registry

we might want to take a look at that as second phase.

**Review by MaysaMacedo — COMMENTED** [review]
In general it looks good to me, great work! Let's ait for CI

File: ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh (hunk 3 above)
Comment (EmilienM) [reply ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh:118]: nice, I just found that it's being used in some steps already:

```
sudo yum install podman openssl -y
curl -L -o mirror-registry.tar.gz https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/mirror-registry/latest/mirror-registry.tar.gz
tar -xzvf mirror-registry.tar.gz
sudo ./mirror-registry install --quayHostname ${OMR_HOST_NAME} --initPassword password --initUser quay -v
```

I'll look at it in a second iteration.

File: ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh (hunk 2 above)
Comment (EmilienM) [reply ci-operator/step-registry/openstack/provision/mirror/openstack-provision-mirror-commands.sh:64]: you're right, let me do this now

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
We need to run pull-ci-openshift-machine-config-operator-master-e2e-openstack-singlestackv6 and pull-ci-openshift-installer-master-e2e-openstack-singlestackv6 as well to test it went fine as it does in periodics.

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6 pull-ci-openshift-machine-config-operator-master-e2e-openstack-singlestackv6 pull-ci-openshift-installer-master-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
/pj-rehearse periodic-ci-shiftstack-ci-release-4.18-e2e-openstack-singlestackv6

**Comment by EmilienM** [comment]
we need to patch MCO again to mount the config drive so we can get what we need to start kubelet.

**Comment by EmilienM** [comment]
Waiting for https://github.com/openshift/machine-config-operator/pull/4586 to land and be promoted in a CI build.

**Comment by EmilienM** [comment]
> /lgtm

thanks.
I'll merge it as is and we submit changes in other PRs.

/pj-rehearse ack
/test build-clusters

**Pull request merged**